internal/cloudsql/refresh_test.go

```diff
@@ -15,77 +15,363 @@
+			desc:    "When the server cert does not decode",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(
+					cn.project, cn.region, cn.name,
+					mock.WithRegion("my-region"),
+					mock.WithCertSigner(func(_ *x509.Certificate, _ *rsa.PrivateKey) ([]byte, error) {
+						certPEM := &bytes.Buffer{}
+						pem.Encode(certPEM, &pem.Block{
+							Type:  "CERTIFICATE",
+							Bytes: []byte("hello"), // woops no cert
+						})
+						return certPEM.Bytes(), nil
+					}),
+				), 1),
+			wantErr: "failed to parse",
+			desc:    "When the cert is not a valid X.509 cert",
+		},
+	}
+	for i, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			client, cleanup, err := mock.NewSQLAdminService(
+				context.Background(),
+				tc.req,
+			)
+			if err != nil {
+				t.Fatalf("failed to create test SQL admin service: %s", err)
+			}
+			defer cleanup()
+
+			r := newRefresher(time.Hour, 30*time.Second, 1, client)
+			_, _, _, err = r.performRefresh(context.Background(), cn, RSAKey)
+
+			if !errorContains(err, tc.wantErr) {
+				t.Errorf("[%v] PerformRefresh failed with unexpected error, want = %v, got = %v", i, tc.wantErr, err)
+			}
+		})
+	}
+}
+
+func TestRefreshWithFailedEphemeralCertCall(t *testing.T) {
+	cn, _ := parseConnName("my-project:my-region:my-instance")
 	inst := mock.NewFakeCSQLInstance(cn.project, cn.region, cn.name)
 
-	// mock expected requests
+	testCases := []struct {
+		reqs    []*mock.Request
+		wantErr string
+		desc    string
+	}{
+		{
+			reqs:    []*mock.Request{mock.InstanceGetSuccess(inst, 1)}, // no ephemeral cert call registered
+			wantErr: "fetch ephemeral cert failed",
+			desc:    "When the CreateEphemeralCert call fails",
+		},
+		{
+			reqs: []*mock.Request{mock.InstanceGetSuccess(inst, 1),
+				mock.CreateEphemeralSuccess(
+					mock.NewFakeCSQLInstance(cn.project, cn.region, cn.name,
+						mock.WithClientCertSigner(
+							func(*x509.Certificate, *rsa.PrivateKey, *rsa.PublicKey) ([]byte, error) {
+								return nil, nil
+							}),
+					), 1),
+			},
+			wantErr: "failed to decode valid PEM cert",
+			desc:    "When decoding the cert fails",
+		},
+		{
+			reqs: []*mock.Request{mock.InstanceGetSuccess(inst, 1),
+				mock.CreateEphemeralSuccess(
+					mock.NewFakeCSQLInstance(cn.project, cn.region, cn.name,
+						mock.WithClientCertSigner(
+							func(*x509.Certificate, *rsa.PrivateKey, *rsa.PublicKey) ([]byte, error) {
+								certPEM := &bytes.Buffer{}
+								pem.Encode(certPEM, &pem.Block{
+									Type:  "CERTIFICATE",
+									Bytes: []byte("hello"), // woops no cert
+								})
+								return certPEM.Bytes(), nil
+							}),
+					), 1),
+			},
+			wantErr: "failed to parse as x509 cert",
+			desc:    "When parsing the cert fails",
+		},
+	}
+	for i, tc := range testCases {
+		mc, url, cleanup := mock.HTTPClient(tc.reqs...)
+		client, err := sqladmin.NewService(
+			context.Background(),
+			option.WithHTTPClient(mc),
+			option.WithEndpoint(url),
+		)
```

Comment:
Should this use `NewSQLAdminService`? 

We may want to just not export HTTPClient anymore, to make sure we are using this new method.

Reply:
Yes. I'll make a pass on all the tests and make sure we're just using `NewSQLAdminService` everywhere.